torch/csrc/api/include/torch/nn/module.h

```diff
@@ -113,25 +113,39 @@ class CloneableModule : public Module {
  public:
   using Module::Module;
 
-  // should it also detach the gradients, like a deepcopy? Or maybe let's just
-  // give clone() a boolean for this?
-  std::unique_ptr<Module> clone() const override {
-    auto ptr = std::unique_ptr<Module>(
-        new Derived(*static_cast<const Derived*>(this)));
+  virtual void reset() = 0;
+
+  std::shared_ptr<Derived> build() {
+    auto module = std::make_shared<Derived>(static_cast<Derived&&>(*this));
+    module->reset();
+    return std::move(module);
+  }
+
+  std::shared_ptr<Module> clone() const override {
+    auto ptr = std::make_shared<Derived>(*static_cast<const Derived*>(this));
+    ptr->parameters_.clear();
+    ptr->reset();
     for (auto& parameter : ptr->parameters_) {
-      parameter.second = this->parameters_.at(parameter.first).clone();
+      parameter.second.data().copy_(
+          this->parameters_.at(parameter.first).data());
     }
     for (auto& child : ptr->children_) {
       child.second = this->children_.at(child.first)->clone();
     }
     return ptr;
   }
 };
-
-template <class Module>
-std::shared_ptr<typename std::decay<Module>::type> make(Module&& module) {
-  auto ptr = std::make_shared<typename std::decay<Module>::type>(
-      std::forward<Module>(module));
-  return ptr;
-}
 }} // namespace torch::nn
+
+#define TORCH_PARAMETER(T, name)             \
+ public:                                     \
+  auto name(T new_##name)->decltype(*this) { \
+    this->name##_ = std::move(new_##name);   \
+    return *this;                            \
+  }                                          \
+  const T& name() const noexcept {           \
+    return this->name##_;                    \
+  }                                          \
+                                             \
+ protected:                                  \
```

Comment:
Wouldn't this randomly change the permissions of the following functions? I think I'd rather just having not do that, it's okay if we leak a few private implementations.

Reply:
I think we shouldn't change visibility inside this macro at all. This will enforce using getters/setters on people, and it might turn out to be quite annoying. Can't we just declare the member and have another macro for setting getters or setters (so they can be possibly used in sections of different visibility, which is controlled by the user)?